File: src/main/java/club/sk1er/mods/core/universal/wrappers/message/UniversalMessage.java
package club.sk1er.mods.core.universal.wrappers.message;

import club.sk1er.mods.core.universal.UniversalMinecraft;
import club.sk1er.mods.core.universal.UniversalPacket;
import club.sk1er.mods.core.universal.wrappers.UniversalPlayer;

//#if FABRIC
//$$ import net.minecraft.text.MutableText;
//#endif

import java.util.ArrayList;
import java.util.List;

/**
 * A UniversalMessage can be thought of as a wrapper around
 * a series of UniversalTextComponents. This class allows
 * functionality such as editing and deleting messages that
 * have already been sent to the user
 */
public class UniversalMessage {
    private UniversalTextComponent chatMessage;
    private List<UniversalTextComponent> messageParts = new ArrayList<>();

    private int chatLineId = -1;
    private boolean recursive = false;
    private boolean formatted = true;

    public UniversalMessage(UniversalTextComponent component) {
        if (component.getSiblings().isEmpty()) {
            messageParts.add(component);
        } else {
            //#if FORGE
            component.getSiblings().stream().map(UniversalTextComponent::new).forEach(messageParts::add);
            //#else
            //$$ component.getSiblings().stream()
            //$$     .map(it -> new UniversalTextComponent((MutableText) it))
            //$$     .forEach(messageParts::add);
            //#endif
        }
    }

    /**
     * Creates a new Message from a list of strings or UniversalTextComponents
     * @param parts
     */
    public UniversalMessage(List<Object> parts) {
        parts.forEach(this::addPart);
    }

    public UniversalTextComponent getChatMessage() {
        parseMessage();
        return chatMessage;
    }

    public String getFormattedText() {
        return getChatMessage().getFormattedText();
    }

    public String getUnformattedText() {
        return getChatMessage().getUnformattedText();
    }

    public List<UniversalTextComponent> getMessageParts() {
        return messageParts;
    }

    public int getChatLineId() {
        return chatLineId;
    }

    public UniversalMessage setChatLineId(int id) {
        chatLineId = id;
        return this;
    }

    public boolean isRecursive() {
        return recursive;
    }

    public UniversalMessage setRecursive(boolean recursive) {
        this.recursive = recursive;
        return this;
    }

    public boolean isFormatted() {
        return formatted;
    }

    public UniversalMessage setFormatted(boolean formatted) {
        this.formatted = formatted;
        return this;
    }

    public UniversalMessage setTextComponent(int index, Object component) {
        if (component instanceof String) {
            messageParts.set(index, new UniversalTextComponent((String) component));
        } else {
            UniversalTextComponent.from(component).ifPresent(it -> messageParts.set(index, it));
        }
        return this;
    }

    public UniversalMessage addTextComponent(int index, Object component) {
        if (component instanceof String) {
            messageParts.add(index, new UniversalTextComponent((String) component));
        } else {
            UniversalTextComponent.from(component).ifPresent(it -> messageParts.add(index, it));
        }
        return this;
    }

    public UniversalMessage addTextComponent(Object component) {
        return addTextComponent(messageParts.size(), component);
    }

    public void edit(UniversalMessage... replacements) {
        throw new UnsupportedOperationException("TODO");
    }

    public void chat() {
        parseMessage();

        if (UniversalPlayer.getPlayer() == null)
            return;

        //TODO: Access transformer
        //#if MC<11602
        if (chatLineId != -1) {
            if (UniversalMinecraft.getChatGUI() != null)
                UniversalMinecraft.getChatGUI().printChatMessageWithOptionalDeletion(chatMessage, chatLineId);
            return;
        }
        //#endif

        if (recursive) {
            UniversalPacket.sendChatMessage(chatMessage);
        } else {
            UniversalPlayer.sendClientSideMessage(chatMessage);
        }
    }

    public void actionBar() {
        parseMessage();

        if (UniversalPlayer.getPlayer() == null)
            return;

        UniversalPacket.sendActionBarMessage(chatMessage);
    }

    private void addPart(Object part) {
        if (part instanceof UniversalTextComponent) {
            messageParts.add((UniversalTextComponent) part);
        } else if (part instanceof String) {
            messageParts.add(new UniversalTextComponent((String) part));
        } else {
            UniversalTextComponent.from(part).ifPresent(this::addPart);
        }
    }

    private void parseMessage() {
        chatMessage = new UniversalTextComponent("");
        messageParts.forEach(chatMessage::appendSibling);
    }
}

File: src/main/kotlin/club/sk1er/mods/core/universal/wrappers/message/UniversalMessage.kt
package club.sk1er.mods.core.universal.wrappers.message

import club.sk1er.mods.core.universal.UniversalPacket
import club.sk1er.mods.core.universal.wrappers.UniversalPlayer
import java.util.function.Function

class UniversalMessage {
    private lateinit var _chatMessage: UniversalTextComponent
    val messageParts = mutableListOf<UniversalTextComponent>()

    val chatMessage: UniversalTextComponent
        get() {
            parseMessage()
            return _chatMessage
        }

    val formattedText: String
        get() = chatMessage.formattedText

    val unformattedText: String
        get() = chatMessage.unformattedText

    var chatLineId: Int = -1
    var isRecursive = false
    var isFormatted = true

    constructor(component: UniversalTextComponent) {
        if (component.siblings.isEmpty()) {
            messageParts.add(component)
        } else {
            //#if FORGE
            component.siblings.map(::UniversalTextComponent).forEach { messageParts.add(it) }
            //#else
            //$$ component.siblings
            //$$         .map { UniversalTextComponent(it as MutableText) }
            //$$         .forEach { messageParts.add(it) }
            //#endif
        }
    }

    constructor(vararg parts: Any) {
        parts.forEach(::addPart)
    }

    fun setTextComponent(index: Int, component: Any) = apply {
        if (component is String) {
            messageParts[index] = UniversalTextComponent(component)
        } else {
            UniversalTextComponent.from(component).ifPresent { messageParts[index] = it }
        }
    }

    fun addTextComponent(index: Int, component: Any) = apply {
        if (component is String) {
            messageParts.add(index, UniversalTextComponent(component))
        } else {
            UniversalTextComponent.from(component).ifPresent { messageParts.add(index, it) }
        }
    }

    fun addTextComponent(component: Any) = addTextComponent(messageParts.size, component)

    fun edit(vararg replacements: UniversalMessage) {
        TODO()
    }

    fun chat() {
        parseMessage()

        if (!UniversalPlayer.hasPlayer())
            return

        // TODO: expose this field in MC>=11602
        //#if MC<11602
        //$$ if (chatLineId != -1) {
        //$$     UniversalMinecraft.getChatGUI()?.printChatMessageWithOptionalDeletion(chatMessage, chatLineId)
        //$$     return
        //$$ }
        //#endif

        if (isRecursive) {
            UniversalPacket.sendChatMessage(_chatMessage)
        } else {
            UniversalPlayer.sendClientSideMessage(_chatMessage)
        }
    }

    fun actionBar() {
        parseMessage()

        if (!UniversalPlayer.hasPlayer())
            return

        UniversalPacket.sendActionBarMessage(_chatMessage)
    }

    private fun addPart(part: Any) {
        when (part) {
            is UniversalTextComponent -> messageParts.add(part)
            is String -> messageParts.add(UniversalTextComponent(part))
            else -> UniversalTextComponent.from(part).ifPresent(::addPart)
        }
    }

    private fun parseMessage() {
        _chatMessage = UniversalTextComponent("")
        messageParts.forEach { _chatMessage.appendSibling(it) }
    }
}
